### PR TITLE
Add TT implementation.

### LINKED ISSUE #17: Implement Transposition Table
Read https://www.chessprogramming.org/Transposition_Table
Transposition Table is a cache table for search.
Implement transposition table which can store 3 types of scores:

1. EXACT - exact value found during search. We should store it every time we found exact value, 
that is when `alpha < score < beta`
2. BETA - should be stored when search fails high, that is when `alpha < beta <= score`
3. ALPHA - should be stored when search fails low, that is when neither 1 or 2 happened. (`score <= alpha`)

Along with scores we should also store:
- depth (on which we write entry to transposition table)
- move associated with found score

If entry already exists for the position, replace it only if `current_depth > old_entry_depth`

Add reading transposition table to the beginning of the search (before we jump into searching moves):
In case we've found cache entry for the position and `depth_entry >= current_depth` then we can return cached score.

Let's make cache size = 16MB for now
Index the table using zobryist hashes (they are implemented in our base  - `chess.hpp`)
`entry = table[position.hash % table_size]`

### PR DESCRIPTION
**Pull request by DomiKoPL**

### CODE REVIEW

**Comment by DomiKoPL** [comment]
I have to add TT to Search. However search currently is just a simple function, so I would need to add it as a parameter. Weird. I have to think if I should do it correctly now or temporarily do it like that.

**Comment by Gustav2137** [comment]
TT should be either a global variable or we can wrap search into some class and put TT as a field there.

**Comment by DomiKoPL** [comment]
Yes, we should. There should be a class that implements Engine like MinMaxEngine or sth.

**Comment by Gustav2137** [comment]
Oh that's what the engine class is for. I can refactor code to fit this template then.

**Comment by Gustav2137** [comment]
I have just committed changes we were discussing to #25. Feel free to review them.

**Comment by Gustav2137** [comment]
> Reference

Ok done.

File: cpp_bestia/src/tt.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+TTData::TTData(int32_t score, Move move, uint8_t depth, uint8_t age, Type type)
+    : data_(pack(score, move, depth, age, type)) {}
+
+int32_t TTData::score() const { return data_ & SCORE_MASK; }
+
+Move TTData::move() const { return (data_ >> MOVE_SHIFT) & MOVE_MASK; }
+
+uint8_t TTData::depth() const { return (data_ >> DEPTH_SHIFT) & DEPTH_MASK; }
+
+uint8_t TTData::age() const { return (data_ >> DEPTH_SHIFT) & DEPTH_MASK; }
```
Comment (Gustav2137) [inline cpp_bestia/src/tt.cpp:16]: Should be AGE_SHIFT and AGE_MASK.

**Comment by Gustav2137** [comment]
lgtm, merging

**Pull request merged**